### ISSUE TITLE
Test SlateDB with DST (deterministic simulation testing)

### ISSUE THREAD

**Issue by brancz** [slatedb/slatedb#267]
**Is your feature request related to a problem? Please describe.**

This isn't a classic feature request, this isn't about a direct feature, and more about creating confidence, that SlateDB indeed delivers on its consistency guarantees (snapshot isolation).

**Describe the solution you'd like**

In my experience, the best way to get started is to write a variety of test cases that represent patterns of workloads against SlateDB, and test that snapshot isolation holds true at all times. I've seen various approaches, but a fairly simple one to get started with is that the simulator plans a set of operations to perform against the database (including failures), let's say 1000 random operations. These operations are seeded with a seed that is generated outside of the simulator so that a test run can be reproduced with the seed. Crucial for reproducibility is that everything is run on a single operating system thread. I don't have first-hand experience with it, but the [madsim](https://github.com/madsim-rs/madsim) project looks interesting, I'm not 100% convinced it's necessary though, since SlateDB has very little dependencies (only object storage). Using the in-memory `object_store` bucket and the tokio [Single-threaded scheduler](https://docs.rs/tokio/latest/tokio/runtime/index.html#current-thread-scheduler) (very importantly without anything, not even any dependencies, ever calling [`spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) should work just fine.

**Describe alternatives you've considered**

There are various ways to implement this, but I think the most pragmatic thing is to start without madsim, and see how that goes, and if things like controlling time is too difficult to retrofit, use madsim. madsim might also be interesting because it looks like it can inject a number of failures. Implementing a simulator from scratch can be a lot of work and may very invasive, but I've seen custom simulators yield the best results. This may still be possible since SlateDB is still young, but in my experience, unless it's designed from the first line of code, it gets harder and harder to retrofit with time.

**Additional context**

There are a few great resources on DST:
* Sled's simulation guide: https://sled.rs/simulation.html
* FoundationDB: https://www.youtube.com/watch?v=4fFDFbi3toc
* TigerBeetle: https://docs.tigerbeetle.com/about/vopr/
* FrostDB: https://www.polarsignals.com/blog/posts/2024/05/28/mostly-dst-in-go

In addition to SlateDB having its simulator, to apply to the [Antithesis](https://antithesis.com/) [open source program](https://antithesis.com/blog/osgp2024).

**Comment by mohnishbasha** [comment]
I would like to explore this and help get started on this. 
I had to write my own test case seperately to test my last pr.

**Comment by criccomini** [comment]
I'm thinking of taking this up. Let me poke around a bit and post some findings over the next week or two.

**Comment by brancz** [comment]
We'll be writing about it more in the new year, but we just finished a first version of our simulator. We closely followed [sled's simulation guide](https://sled.rs/simulation.html). To do this we turned our entire system into a set of finite-state machines. Happy to share our design doc for it privately (not quite ready to talk about it publicly since it also reveals some details about the system we're building, but happy to share privately). FWIW even for our new system, this took us a good 3 weeks to do, so this is no insignificant amount of work.

**Comment by criccomini** [comment]
I'm investigating what work needs to be done if we want to do DST without any 3rd party stuff. Some [things to look at](https://s2.dev/blog/dst):

1. Execution – Single-threaded to eliminate scheduler noise
1. Entropy – All RNGs should have a known seed
1. Time – No physical clocks
1. I/O – No dependency on any external IO with components not part of the simulation, and subject to all the same constraints

For (1), we spawn threads in three places:

- [x] The async runtime (tokio)
- [x] The garbage collector (`GarbageCollector::start_in_bg_thread`)
- [x] The compactor (`Compactor::new`).

Limiting the async runtime to a single thread is easy enough. For the GC and compactor, I _think_ we're OK since we can pass the same single-threaded tokio handle into both. In both cases, the threads are just calling `tokio_handle.block_on`, so everything should still go through the handle's single thread. Would love confirmation if my understanding is correct.

_EDIT: It seems like there might be a subtle race condition with the GC/compactor threads. Since we spawn a thread in each and /then/ call spawn blocking, the (non-deterministic) OS scheduler is the one that actually decides whether the GC or compactor block_on call goes first. I think we can add a `RunMode` to the settings to determine whether to directly use the async runtime or spawn a thread._

✅ #620, #632

For (2), I grepped through the codebase, and it looks like the `slatedb` crate has RNGs in...

- [x] `storage_fs.rs` uses `Rng`.
- [x] We use `ULID`s and `Uuid`s in a bunch of places.
- [x] A bunch of tests and test utils use `Rng`. (no changes needed)

It looks like ULID and Uuids both let us inject a seeded generator, so we should be able to use a single RNG for the whole system and then seed it.

✅  #585, #633

For (3)...

- [x] `checkpoint.rs` uses `SystemTime::now()`
- [x] `config.rs`'s `SystemClock`, which is used for TTL and checkpoint times.
- [x] `CompactorEventHandler::finish_compaction` uses `SystemTime::now()` for logging last compaction timestamp in stats.
- [ ] In `moka.rs` if `ttl` or `tti` are set. (Note: it appears Foyer only supports max capacity)
- [x] `CompactorOrchestrator` uses a non-async crossbeam ticker, which uses `std::time::Interval`.
- [x] `db_cache/mod.rs` uses `Instant::now()` when setting `last_err_log_instant`
- [x] `garbage_collector.rs` uses `Utc::now()` in `run_gc_task`.
- [x] `storage_fs.rs::background_evict` uses `SystemTime::now()`

IMO, we should switch all of these over to our `Clock` trait

I am marking the clock work (hah) as done. The Moka TTL/TTI stuff is still not mockable, but I opened https://github.com/moka-rs/moka/issues/522. In the meantime, we can do DST with a max capacity rather than TTL/TTI for the cache.

✅  #603, #601, #620

For (4), I think we have:

- [ ] I/O with the disk (for local filesystem caching)
- [x] The ObjectStore crate

In the case of the object stores, I think we can use `InMemory` for everything. Probably worth checking if the hashing it uses for its internal map is deterministic. For disk, we're using `tokio::fs`, so it might be OK, I'm not sure. If not, we'd have to mock it.

_EDIT: For `tokio::fs`, it looks like we can use [`sos_vfs`](https://docs.rs/sos-vfs/0.2.6/sos_vfs/). This crate maps directly to `tokio::fs`, so we can add a feature in `storage_fs.rs` to enable `sos_vfs` when we want to run the DST. This mock doesn't include fault injection, but it at least makes the FS deterministic (I checked that it's using BTreeMap for in-memory directory iteration)._

Everything seems fairly doable, just a matter of slogging through it.

**Comment by criccomini** [comment]
I spent the past two days investigating sos-vfs. Unfortunately, it doesn't work.

It's very buggy. I opened a few issues:

- https://github.com/saveoursecrets/sdk/issues/761
- https://github.com/saveoursecrets/sdk/issues/762

Even with these patches, it failed on other unit tests that pass just fine with `tokio::fs`. I've already burned way too much time on this, so I'm going to back off on the filesystem mocking for the moment.

**Comment by criccomini** [comment]
@dmvk Pointed me at https://docs.rs/loom/latest/loom/ as well. It seems complementary to the DST I had in mind. My plan was (and is) to mock out all non-determinism and use a single Tokio threaded runtime to control scheduling. With Loom, the scheduling is handed over to its runtime, which uses state reduction to explore the various code paths.

tl;dr I'm going to scope this GH issue to implementing DST with a single threaded Tokio runtime. It'll be less efficient since it won't do state reduction, but I suspect it'll still surface a lot of bugs. Separately, we can have a separate issue to run the same DST tests in the Loom runtime.
